File: benchmarks/gated_vs_ungated_resonators.py
#!/usr/bin/env python3
"""
Benchmark: amplitude-gated Resonator energy update vs the original
always-compute version, at scale.

Claim being tested: skipping the expensive `x*x` multiply when a
resonator's amplitude is below the noise floor (see runtime.py's
ResonatorState.step docstring) saves real computation in a bank where
most channels are quiet most of the time -- the same dormancy idea
proven for NPC brains in dormant_vs_polling_heavy.c, applied to
resonator channels instead.

Setup: a bank of M resonator channels (e.g. a multi-band audio/sensor
analyzer), only a small number of which are ever "loud" (driven near
their resonance) at once -- everything else sees pure noise. Both
versions integrate the SAME oscillator mechanics every tick (x, v always
update -- a real signal could arrive on any channel at any time, so that
part can't be skipped). The only difference is whether the energy
update's multiply is gated.

Run it:
    python gated_vs_ungated_resonators.py
"""

import math
import random
import time
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "core"))
from runtime.runtime import ResonatorState, DEFAULT_RESONATOR_BASE_GAIN


def make_bank(channel_freqs):
    bank = []
    for f in channel_freqs:
        omega = 2 * math.pi * f
        coupling = DEFAULT_RESONATOR_BASE_GAIN * (omega ** 2)
        bank.append(ResonatorState(name=f"R{f}", freq_hz=f, damping=0.02, coupling=coupling))
    return bank


def step_ungated(r: ResonatorState, drive: float, dt: float):
    """Same math as ResonatorState.step but with gating forced off --
    i.e. the ORIGINAL always-compute behavior, for comparison."""
    omega = r.omega
    accel = -(omega ** 2) * r.x - 2 * r.damping * omega * r.v
    accel += r.coupling * drive
    r.v += accel * dt
    r.x += r.v * dt
    r.energy_ema += 0.01 * (r.x * r.x - r.energy_ema)  # always pays the multiply


def run_trial(n_channels, loud_fraction, ticks, gated: bool):
    freqs = [110 + i * 30 for i in range(n_channels)]  # spread across a band
    bank = make_bank(freqs)
    n_loud = max(1, int(n_channels * loud_fraction))
    loud_idx = set(random.sample(range(n_channels), n_loud))

    dt = 1.0 / 40000

    start = time.perf_counter()
    for t in range(ticks):
        time_s = t * dt
        for i, r in enumerate(bank):
            if i in loud_idx:
                drive = 1.0 * math.sin(2 * math.pi * r.freq_hz * time_s) + random.uniform(-0.05, 0.05)
            else:
                drive = random.uniform(-0.05, 0.05)  # noise only -- this channel stays quiet
            if gated:
                r.step(drive, dt)
            else:
                step_ungated(r, drive, dt)
    return time.perf_counter() - start


def main():
    ticks = 3000
    channel_counts = [16, 64, 256]
    loud_fractions = [0.02, 0.1, 0.3, 1.0]

    print(f"{'Channels':>9} {'%loud':>7} {'Ungated (s)':>13} {'Gated (s)':>11} {'Speedup':>9}")
    print("-" * 56)
    for n in channel_counts:
        for frac in loud_fractions:
            random.seed(11)
            t_ungated = run_trial(n, frac, ticks, gated=False)
            random.seed(11)
            t_gated = run_trial(n, frac, ticks, gated=True)
            speedup = t_ungated / t_gated if t_gated > 0 else float("inf")
            print(f"{n:>9} {frac*100:>6.0f}% {t_ungated:>13.4f} {t_gated:>11.4f} {speedup:>8.2f}x")
        print()


if __name__ == "__main__":
    main()

File: benchmarks/gated_vs_ungated_resonators.c
/*
 * C version of gated_vs_ungated_resonators.py -- tests whether
 * amplitude-gating the resonator energy update (skip x*x when |x| is
 * below the noise floor) is actually faster in compiled code, or
 * whether (like benchmark #1/#2's lesson) the gate's branch costs more
 * than the multiply it avoids -- especially relevant here since modern
 * CPUs pipeline multiplies cheaply but pay real misprediction penalties
 * on data-dependent (unpredictable) branches, which "which channel is
 * loud" inherently is.
 *
 * Build: gcc -O2 -o gated_vs_ungated_resonators.exe gated_vs_ungated_resonators.c -lm
 * Run:   ./gated_vs_ungated_resonators.exe
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

typedef struct {
    double freq_hz, damping, coupling, threshold, gate_threshold;
    double x, v, energy_ema;
} Resonator;

static void init_resonator(Resonator *r, double freq_hz) {
    double omega = 2.0 * M_PI * freq_hz;
    r->freq_hz = freq_hz;
    r->damping = 0.02;
    r->coupling = 4.0e-4 * omega * omega;
    r->threshold = 0.0008;
    r->gate_threshold = 0.00024;
    r->x = 0.0; r->v = 0.0; r->energy_ema = 0.0;
}

static void step_ungated(Resonator *r, double drive, double dt) {
    double omega = 2.0 * M_PI * r->freq_hz;
    double accel = -(omega * omega) * r->x - 2.0 * r->damping * omega * r->v;
    accel += r->coupling * drive;
    r->v += accel * dt;
    r->x += r->v * dt;
    r->energy_ema += 0.01 * (r->x * r->x - r->energy_ema);  // always pays the multiply
}

static void step_gated(Resonator *r, double drive, double dt) {
    double omega = 2.0 * M_PI * r->freq_hz;
    double accel = -(omega * omega) * r->x - 2.0 * r->damping * omega * r->v;
    accel += r->coupling * drive;
    r->v += accel * dt;
    r->x += r->v * dt;
    if (fabs(r->x) >= r->gate_threshold) {
        r->energy_ema += 0.01 * (r->x * r->x - r->energy_ema);
    } else {
        r->energy_ema *= 0.99;
    }
}

static double frand(void) { return (double)rand() / RAND_MAX; }

static double run_trial(int n_channels, double loud_fraction, int ticks, int gated) {
    Resonator *bank = malloc(sizeof(Resonator) * n_channels);
    for (int i = 0; i < n_channels; i++) init_resonator(&bank[i], 110.0 + i * 30.0);

    int n_loud = (int)(n_channels * loud_fraction);
    if (n_loud < 1) n_loud = 1;
    int *is_loud = calloc(n_channels, sizeof(int));
    for (int k = 0; k < n_loud; k++) is_loud[k] = 1; /* first n_loud channels are loud -- order doesn't matter, distribution is what's tested */

    double dt = 1.0 / 40000.0;
    clock_t start = clock();
    for (int t = 0; t < ticks; t++) {
        double time_s = t * dt;
        for (int i = 0; i < n_channels; i++) {
            double drive;
            if (is_loud[i]) {
                drive = sin(2.0 * M_PI * bank[i].freq_hz * time_s) + (frand() - 0.5) * 0.1;
            } else {
                drive = (frand() - 0.5) * 0.1;
            }
            if (gated) step_gated(&bank[i], drive, dt);
            else step_ungated(&bank[i], drive, dt);
        }
    }
    double elapsed = (double)(clock() - start) / CLOCKS_PER_SEC;
    free(bank); free(is_loud);
    return elapsed;
}

int main(void) {
    srand(11);
    int ticks = 20000;
    int channel_counts[] = {16, 64, 256};
    double loud_fractions[] = {0.02, 0.10, 0.30, 1.00};

    printf("%9s %7s %13s %11s %9s\n", "Channels", "%loud", "Ungated (s)", "Gated (s)", "Speedup");
    for (size_t ci = 0; ci < sizeof(channel_counts)/sizeof(int); ci++) {
        for (size_t fi = 0; fi < sizeof(loud_fractions)/sizeof(double); fi++) {
            int n = channel_counts[ci];
            double f = loud_fractions[fi];
            srand(11);
            double t_ungated = run_trial(n, f, ticks, 0);
            srand(11);
            double t_gated = run_trial(n, f, ticks, 1);
            double speedup = t_gated > 0 ? t_ungated / t_gated : 0.0;
            printf("%9d %6.0f%% %13.4f %11.4f %8.2fx\n", n, f * 100, t_ungated, t_gated, speedup);
        }
        printf("\n");
    }
    return 0;
}
